### ISSUE TITLE
DynamicalDecoupling does not support non-basis gates

### ISSUE THREAD

**Issue by nbronn** [Qiskit/qiskit-terra#7400]
### Environment

- **Qiskit Terra version**: 0.20.0.dev0+309bfb1
- **Python version**: 3.9.4
- **Operating system**: Darwin

### What is happening?

When generalizing [this tutorial](https://qiskit.org/documentation/stubs/qiskit.transpiler.passes.DynamicalDecoupling.html]) on using the `DynamicalDecoupling` transpiler pass, including other sequences involving gates that are not native to the backend (i.e., `YGate()`) fail when run on an IBM backend (or transpiled and scheduled for one). Two solutions are presented here that attempt to address this: using the `.to_gate()` method to construct a `Gate` from a `QuantumCircuit` that is effectively a `YGate()` (which fails), and using `Pulse` to create a `schedule` that can be added as a calibration to the `'y'` gate.

### How can we reproduce the issue?

```
qc = QuantumCircuit(4)
qc.h(0)
qc.cx(0, 1)
qc.cx(1, 2)
qc.cx(2, 3)
qc.measure_all()
qc_trans = transpile(qc, backend, initial_layout=[0, 1, 3, 4]) # ibmq_belem

dd_XY4_sequence = [XGate(), YGate(), XGate(), YGate()]
pm = PassManager([ALAPSchedule(durations),
                  DynamicalDecoupling(durations, dd_XY4_sequence)])
qc_XY4_sequence = pm.run(qc_trans)

schedule(qc_XY4_sequence, backend).draw()
```
results in
```
QiskitError: "Operation 'y' on qubit(s) [0] not supported by the backend command definition. Did you remember to transpile your input circuit for the same backend?"
```

Similarly, attempting to send the job to the backend via
```
job = backend.run(qc_XY4_sequence, meas_level=2, shots=8192) 
```
results in the following IQX error
```
The Qobj uses gates (['y']) that are not among the basis gates (['id', 'rz', 'sx', 'x', 'cx', 'reset']) [1106]
```

### What should happen?

The `DynamicalDecoupling` would nominally add Pulse schedules for non-basis gates I think (the following code below shows how).

### Any suggestions?

The first attempted solution (which does not work), creates a `Gate` from an equivalent `QuantumCircuit`:
```
my_y_circ = QuantumCircuit(1)
my_y_circ.s(0)
my_y_circ.x(0)
my_y_circ.sdg(0)
my_y_gate = my_y_circ.to_gate()
```
checking that this `Gate` is equivalent to the `YGate()` provided by the circuit library,
```
from qiskit.quantum_info import Operator
Operator(YGate()).equiv(Operator(my_y_gate))
```
returns
```
True
```
However, now
```
dd_XY4_sequence = [XGate(), my_y_gate, XGate(), my_y_gate]
pm = PassManager([ALAPSchedule(durations),
                  DynamicalDecoupling(durations, dd_XY4_sequence)])
qc_XY4_sequence = pm.run(qc_trans)
```
returns
```
CircuitError: "to_matrix not defined for this <class 'qiskit.circuit.gate.Gate'>"
```
and there does not appear to be a method to add a matrix representation to a `Gate`.

The second attempt works but requires a detailed knowledge of `Pulse`, which seems like an unnecessary hurdle for a user simply wanting to implement dynamical decoupling (this includes detailed knowledge of how the backend works as well). First, when calculating durations, let the transpiler passes know that a `'y'` gate duration is the same as `'x'`:
```
inst_durs = []

# single qubit gates
for qubit in range(backend.config().num_qubits):
    for inst_str in inst_sched_map.qubit_instructions(qubits=[qubit]):
        inst = inst_sched_map.get(inst_str, qubits=[qubit])
        inst_durs.append((inst_str, qubit, inst.duration))
        if inst_str == 'x':
            inst_durs.append(('y', qubit, inst.duration))
        
# two qubit gates
for qc in range(backend.config().num_qubits):
    for qt in range(backend.config().num_qubits):
        for inst_str in inst_sched_map.qubit_instructions(qubits=[qc, qt]):
            inst = inst_sched_map.get(inst_str, qubits=[qc, qt])
            inst_durs.append((inst_str, [qc, qt], inst.duration))
            
durations = InstructionDurations(inst_durs)
```
Next, create a `Schedule` in the `Pulse` builder that implements the `'y'` gate in terms of the `'x'` gate, and add it as a calibration:
```
for qubit in range(backend.config().num_qubits):
    with pulse.build('y gate for qubit '+str(qubit)) as sched:
        with pulse.phase_offset(np.pi/2, DriveChannel(qubit)):
            x_gate = inst_sched_map.get('x', qubits=[qubit])
            pulse.call(x_gate)
    
        qc_XY4_sequence.add_calibration('y', [qubit], sched)
```
which can be scheduled and run on an IQX backend.

Nominally I think the latter approach should be added for `Gate`s in Qiskit's circuit library (or at least the single qubit ones) inside the `DynamicalDecoupling` pass (or another one that explicitly does this, similar to `RZXCalibrationBuilder`). In general I think functionality that allows a user to specify the matrix representation of what they wish to achieve to the gate itself (such as with `PulseGates`). A similar issue to this has been noted with respect to interleaved randomized benchmarking in `Experiment`, where the user cannot specify the desired Clifford element of `PulseGate`s.

**Comment by jakelishman** [comment]
I think it's expected that the code in "how can we reproduce the error" should fail; it explicitly tries to use a `YGate` on a backend that doesn't support it.  The general fix would be to make it easier to pass the dynamical decoupling sequence through the basis translator, but I'm not really sure that makes any physical sense; surely your choice of dynamical decoupling is going to _strongly_ depend on the available gate set of the hardware backend?

**Comment by nbronn** [comment]
I think the interesting thing here is that Qiskit knows that the `YGate()` used in the DD sequence here corresponds to the `'y'` instruction I added to the `instruction_schedule_map` (as also in the error received in "How we can reproduce the issue?"). 

Of course that's correct that your choice of DD sequence is going to depend on the available gate set, and I think making a transpiler pass that can make sense of easily-expressible gates for DD in terms of pulse gates would be valuable (as with the calibration builders). For example, a simpler case is the `Xp-Xm` DD sequence IBM uses in experiments such as [QV 64](https://arxiv.org/abs/2008.08571) and [scalable error mitigation](https://arxiv.org/abs/2108.09197) is not easily implementable because this is not exactly equivalent to the `[XGate() XGate()]` in the [DD tutorial](https://qiskit.org/documentation/stubs/qiskit.transpiler.passes.DynamicalDecoupling.html). The `Xp` corresponds to the `'x'` instruction, but the `Xm` would be the `'x'` instruction with negative amplitude, which could compensate for amplitude calibration error. The point of this issue is that this should be more easily done in Qiskit.

To summarize, what I would like to do is:
- create a simple DD sequence such as `[XpGate() XmGate()]`
- such that I can define the `.to_matrix()` function corresponding to their intended unitary operation
- that I can put through a transpiler pass that would attach the appropriate pulse schedule derived from the `instruction_schedule_map` (as with calibrations)

From this it should be apparent how to proceed with `YGate()`s and so on.

**Comment by jakelishman** [comment]
A little update: I brought up dynamical decoupling again in an open Terra meeting today, essentially asking for comments on your needs and how we can / if we should implement them.  We agreed to move further discussion here.
